src/Windows/R/Components/Impl/InteractiveWorkflow/Implementation/InteractiveWindowWriter.cs

```diff
@@ -0,0 +1,198 @@
+                _interactiveWindow.Write(" ");
+                // Must flush so we do get 'buffer changed' immediately.
+                _interactiveWindow.FlushOutput();
+            } else {
+                // Locate last end of line
+                snapshot = _interactiveWindow.OutputBuffer.CurrentSnapshot;
+                var line = snapshot.GetLineFromPosition(snapshot.Length);
+                _currentLineStart = line.Start;
+
+            }
+        }
+
+        private void OnBufferChanged(object sender, TextContentChangedEventArgs e) {
+            _interactiveWindow.OutputBuffer.Changed -= OnBufferChanged;
+
+            Debug.Assert(_currentLineStart >= 0);
+            if (_currentLineStart >= 0) {
+                var snapshot = _interactiveWindow.OutputBuffer.CurrentSnapshot;
+
+                var span = Span.FromBounds(_currentLineStart, snapshot.Length);
+                var currentText = snapshot.GetText(span);
+
+                var index = GetDifference(currentText, _newText, out string difference);
+                if (index >= 0) {
+                    span = Span.FromBounds(_currentLineStart + index, snapshot.Length);
+                    _interactiveWindow.OutputBuffer.Replace(span, difference);
+                } else {
+                    _interactiveWindow.OutputBuffer.Replace(span, _newText);
+                }
+            }
+        }
+
+        private static int GetDifference(string oldText, string newText, out string difference) {
+            difference = string.Empty;
+            if (newText.Length < oldText.Length) {
+                return -1;
+            }
+
+            var index = -1;
+
+            for (var i = 0; i < oldText.Length; i++) {
+                if (oldText[i] != newText[i]) {
+                    index = i;
+                    break;
+                }
+            }
+
+            difference = newText.Substring(index);
+            return index;
+        }
+
+        internal class Message {
+            public string Text;
+            public bool IsError;
+
+            public bool MovesCaretBack() => Text.Length > 1 && Text[0] == '\r' && Text[1] != '\n';
+            public bool ContainsLineFeed() => Text.Any(ch => ch == '\n');
+            public bool IsPlainText() => !MovesCaretBack() && !ContainsLineFeed();
+        }
+
+        internal class MessageQueue : IDisposable {
+            private readonly List<Message> _messages = new List<Message>();
+            private readonly ManualResetEventSlim _messagesAvailable = new ManualResetEventSlim(false);
+            private readonly object _lock = new object();
+            private int _lastCRMessageIndex = -1;
+
+            public void Enqueue(string text, bool isError) {
+                lock (_lock) {
+                    var m = new Message { Text = text, IsError = isError };
+                    // Try merging text with the last CR message, if any
+                    if (_messages.Count > 0) {
```

Comment:
I'm not entirely sure, but it looks like this merge can be moved into consumer, and in that case you can use `BufferBlock<Message>` instead of custom queue.